File: typescript/src/sql.ts
/**
 * Query field list for SQL SELECT.
 */
export const QUERY_FIELDS =
  "subject, predicate, object, meta";

/**
 * Convert $pattern arguments to PostgreSQL jsonb containment triple.
 * - ["$pattern", "$*", "author of", "$*"] -> triple: ["author of"]
 * - ["$pattern", "Liu Xin", "author of", "$*"] -> triple: ["Liu Xin", "author of", "$*"]
 */
export function patternToTriple(
  subject: string,
  predicate: string,
  object: string
): unknown[] {
  let pattern = [];
  if (subject !== "*") {
    pattern.push(subject);
  }
  if (predicate !== "*") {
    pattern.push(predicate);
  }
  if (object !== "*") {
    pattern.push(object);
  }
  return pattern;
}

/**
 * Build SQL WHERE clause for a triple pattern.
 */
export function tripleToSqlCondition(triple: unknown[]): string {
  const json = JSON.stringify({ triple });
  const escaped = json.replace(/'/g, "''");
  return `meta @> '${escaped}'`;
}

File: java/src/main/java/io/github/marchliu/jse/Sql.java
package io.github.marchliu.jse;

import java.util.List;

/**
 * SQL helpers for JSE query expressions.
 *
 * <p>This class now delegates to {@link io.github.marchliu.jse.functors.SqlFunctors}
 * for the actual implementation.</p>
 *
 * @deprecated Use {@link io.github.marchliu.jse.functors.SqlFunctors} directly.
 */
@Deprecated
public final class Sql {
    private Sql() {}

    /** Query field list for SQL SELECT. */
    public static final String QUERY_FIELDS = io.github.marchliu.jse.functors.SqlFunctors.QUERY_FIELDS;

    /**
     * Convert $pattern arguments to PostgreSQL jsonb containment triple.
     * <ul>
     *   <li>["*", "author of", "*"] -> ["author of"]</li>
     *   <li>["Liu Xin", "author of", "*"] -> ["Liu Xin", "author of", "*"]</li>
     * </ul>
     *
     * @param subject   Subject pattern
     * @param predicate Predicate pattern
     * @param object    Object pattern
     * @return List of non-wildcard elements
     * @deprecated Use {@link io.github.marchliu.jse.functors.SqlFunctors#patternToTriple(String, String, String)}
     */
    @Deprecated
    public static List<String> patternToTriple(String subject, String predicate, String object) {
        return io.github.marchliu.jse.functors.SqlFunctors.patternToTriple(subject, predicate, object);
    }

    /**
     * Build SQL WHERE clause for a triple pattern.
     *
     * @param triple Triple list
     * @return SQL WHERE condition
     * @deprecated Use {@link io.github.marchliu.jse.functors.SqlFunctors#tripleToSqlCondition(List)}
     */
    @Deprecated
    public static String tripleToSqlCondition(List<String> triple) {
        return io.github.marchliu.jse.functors.SqlFunctors.tripleToSqlCondition(triple);
    }
}
